plugin/src/main/java/org/opensearch/ml/model/MLModelAutoReDeployer.java

```diff
@@ -0,0 +1,471 @@
+    public MLModelAutoReDeployer(
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        Client client,
+        Settings settings,
+        NodeEnvironment nodeEnvironment,
+        NamedXContentRegistry xContentRegistry
+    ) {
+        this.client = client;
+        this.threadPool = threadPool;
+        this.settings = settings;
+        this.nodeEnvironment = nodeEnvironment;
+        this.xContentRegistry = xContentRegistry;
+
+        enableAutoReDeployModel = ML_COMMONS_MODEL_AUTO_REDEPLOY_ENABLE.get(settings);
+        onlyRunOnMlNode = ML_COMMONS_ONLY_RUN_ON_ML_NODE.get(settings);
+        autoDeployMaxRetryTimes = ML_MODEL_REDEPLOY_MAX_RETRY_TIMES.get(settings);
+
+        clusterService
+            .getClusterSettings()
+            .addSettingsUpdateConsumer(ML_COMMONS_MODEL_AUTO_REDEPLOY_ENABLE, it -> enableAutoReDeployModel = it);
+
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(ML_COMMONS_ONLY_RUN_ON_ML_NODE, it -> onlyRunOnMlNode = it);
+
+        clusterService
+            .getClusterSettings()
+            .addSettingsUpdateConsumer(ML_MODEL_REDEPLOY_MAX_RETRY_TIMES, it -> autoDeployMaxRetryTimes = it);
+    }
+
+    /** the main method: model auto redeploying */
+    public void autoReDeployModel() {
+        log.info("auto redeploy model enabled: {} ", enableAutoReDeployModel);
+
+        // if we don't need to redeploy automatically, just return without doing anything
+        if (!enableAutoReDeployModel) {
+            return;
+        }
+
+        String localNodeId = nodeEnvironment.nodeId();
+        // if the local node are both ml and data node,it should throw MLException
+        if (isNode(ML_ROLE_NAME) && isNode(DATA_ROLE_NAME)) {
+            throw new MLException("the node " + localNodeId + " shouldn't be both ml and data node");
+        }
+
+        // auto redeploy all models of this local node
+        threadPool.executor(DEPLOY_THREAD_POOL).execute(() -> {
+            long sleepTimeUnit = 1000L;
+            int numRetries = 4;
+            int numTries = 0;
+
+            while (numTries < numRetries) {
+                ++numTries;
+                log.info("trying to auto-redeploy model. try {} of {}", numTries, numRetries);
+
+                try {
+                    TimeUnit.MILLISECONDS.sleep(sleepTimeUnit * numTries);
+                    autoReDeployModel(localNodeId);
+                    break;
+                } catch (InterruptedException | RuntimeException e) {
+                    log.error("exception while auto-redeploying model.", e);
+                }
+            }
+        });
+    }
+
+    /**
+     * auto re-deploy model under the local node id<br>
+     * @param localNodeId local node id
+     * @throws MLException exception
+     */
+    @VisibleForTesting
+    void autoReDeployModel(String localNodeId) throws MLException {
+        // wait the index ".plugins-ml-model" finish initialization and get all the models to be deployed
```

Comment:
Seems this comment should be moved to before line 147?

Reply:
moved